New welcome flow for default search

imptrx opened the pull request:
- Change flow for how users select their default search engine. Gives user ability to use immediate dropdown rather than go through settings
- See https://github.com/brave/brave-ui/pull/472 for storyboard and new SVG change
- https://github.com/brave/brave-browser/issues/1548
## Submitter Checklist:
- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests && npm run test-security`) on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection

## Test Plan:
- Ensure dropdown works
- Ensure button is disabled if no search engine is selected 
- Ensure display text changes based on locale
- (default) should be appended to the region's default search provider
- Ensure 'Confirm' button still links to next page of on-boarding flow
- Ensure responsiveness on smaller screens
- Ensure new search engine selected becomes the default search engine in address bad

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

cezaraugusto commented on components/brave_welcome_ui/components/screens/searchBox.tsx:
```diff
@@ -21,11 +21,33 @@ interface Props {
-    const { index, currentScreen, onClick } = this.props
+    const { index, currentScreen, onClick, isDefaultSearchGoogle, searchProviders } = this.props
+    const { searchEngineSelected } = this.state
+    const bodyText = isDefaultSearchGoogle ? `${getLocale('chooseSearchEngine')} ${getLocale('privateExperience')}` : getLocale('chooseSearchEngine')
```
please add the locale for `privateExperience`. welcome page locale can be found in:

https://github.com/brave/brave-core/blob/master/browser/ui/webui/brave_webui_source.cc#L174
https://github.com/brave/brave-core/blob/master/components/resources/brave_components_strings.grd#L172

cezaraugusto commented on components/brave_welcome_ui/components/screens/searchBox.tsx:
```diff
@@ -35,14 +57,28 @@ export default class SearchEngineBox extends React.PureComponent<Props, {}> {
+        <Paragraph>{bodyText}</Paragraph>
+          <SelectGrid>
+            <SelectBox onChange={this.onChangeDefaultSearchEngine}>
+              <option key={0} value=''>{getLocale('selectSearchEngine')}</option>
```
ditto

cezaraugusto commented on components/brave_welcome_ui/components/screens/searchBox.tsx:
```diff
@@ -35,14 +57,28 @@ export default class SearchEngineBox extends React.PureComponent<Props, {}> {
+              level='primary'
+              type='accent'
+              size='large'
+              text={getLocale('search')}
```
I guess this should be `Set default` instead?

imptrx commented on components/brave_welcome_ui/components/screens/searchBox.tsx:
```diff
@@ -35,14 +57,28 @@ export default class SearchEngineBox extends React.PureComponent<Props, {}> {
+                  key={index + 1}
+                  value={provider.value}
+                >
+                  {getLocale(provider.name)}
```
@cezaraugusto since this list is populated based on API returns should I just leave this as `provider.name` rather than `getLocale(provider.name)`

cezaraugusto replied:
we can remove `getLocale` here since we don't translate their names

cezaraugusto commented:
thanks @imptrx! front-end is good to go

bsclifton commented:
Will start checking out the back end parts today 😄

bsclifton commented on browser/ui/webui/brave_welcome_ui.cc:
```diff
@@ -67,8 +68,15 @@ BraveWelcomeUI::BraveWelcomeUI(content::WebUI* web_ui, const std::string& name)
   auto handler_owner = std::make_unique<WelcomeDOMHandler>();
   WelcomeDOMHandler* handler = handler_owner.get();
```
NOTE: with a little work (definitely in another PR), we could remove this object in favor of `ImportDataHandler` (in `"chrome/browser/ui/webui/settings/settings_import_data_handler.h"`). On page load, we can call the `initializeImportDialog` method to get all the browser profiles and then let the user choose via a drop down (and when ready, call `chrome.send('importData', [indexChosenFromDropDown)`)

bsclifton commented on components/brave_welcome_ui/components/screens/searchBox.tsx:
```diff
@@ -18,11 +19,64 @@ interface Props {
+  searchProviders: Array<any>
+}
+
+interface SearchEngineEntry {
```
Example entry (returned from the `SearchEnginesHandler`):
```json
{
  "canBeDefault": true,
  "canBeEdited": true,
  "canBeRemoved": true,
  "default": false,
  "displayName": "DuckDuckGo",
  "iconURL": "https://duckduckgo.com/favicon.ico",
  "id": 3,
  "isOmniboxExtension": false,
  "keyword": ":d",
  "modelIndex": 1,
  "name": "DuckDuckGo",
  "url": "https://duckduckgo.com/?q=%s&t=brave",
  "urlLocked": true
}
```

cezaraugusto replied:
@imptrx could you move this interface to welcome.d.ts file?

bsclifton commented on components/brave_welcome_ui/components/screens/searchBox.tsx:
```diff
@@ -18,11 +19,64 @@ interface Props {
+    }
+
+    // @ts-ignore
+    window.cr.sendWithPromise('getSearchEnginesList').then(response => {
```
Not sure how to properly expose this via typescript? so had to ts-ignore it

Method is defined in https://github.com/brave/brave-core/blob/c90dbb4a7166436b1dcb1b3eacbb6184efb8f6a4/components/brave_welcome_ui/brave_welcome.html#L10

cezaraugusto replied:
@imptrx could you create a helper method for this? also add a `catch` at the promise's end with a warning like "unable to get the search engines list" or something you think fits better

cezaraugusto replied:
also please type the method

imptrx replied:
@cezaraugusto I am thinking it be cleaner if we move this method under the welcome page component's `initialize` method and pipe it through redux before populating it here. It looks like we are doing something along those lines in 'brave_adblock` - would that be a good reference point?

cezaraugusto replied:
sounds good 👍

bsclifton commented:
@rebron @imptrx @cezaraugusto let's meet up on this one 😄 

Because of how I implemented the back-end (it sets it immediately), the button next to the drop down likely isn't needed (at the moment, it does the same thing as hitting next, but that's all it does)

simonhong commented on browser/ui/webui/brave_welcome_ui.cc:
```diff
@@ -67,8 +68,15 @@ BraveWelcomeUI::BraveWelcomeUI(content::WebUI* web_ui, const std::string& name)
+  auto search_handler_owner = std::make_unique<settings::SearchEnginesHandler>(profile);
+  web_ui->AddMessageHandler(std::move(search_handler_owner));
+
+  handler->Init();
```
Just a nit:  `Init()` and `OnWalletInitialized()` of `WelcomeDOMHandler` can be deleted.
`Init()` is empty method and the other doesn't have definition.

bsclifton replied:
Done! 👍

cezaraugusto commented:
@rossmoody thoughts on https://github.com/brave/brave-core/pull/2479#issuecomment-499775913?

rossmoody commented:
I built this branch and I actually quite like the functionality we have going on here. I think it's important we keep the button, though. Both cause it signals to the user to make a decision by being disabled but also because it gives them a place to feel like they "confirmed" the choice (even if technically the setting is in place already) instead of advancing what might seem abruptly.

bsclifton commented:
Awesome, thanks @rossmoody 😄I guess this is ready for official review! Did you want to be one of the lucky people to hit the review button?

bsclifton commented:
OK should be good to go!  😄 Review away (CI running... all lints have been fixed - C++/JavaScript)

cezaraugusto commented on components/brave_welcome_ui/components/screens/searchBox.tsx:
```diff
@@ -18,11 +19,62 @@ interface Props {
+
+  onChangeDefaultSearchEngine = (event: React.ChangeEvent<HTMLSelectElement>) => {
+    const modelIndex = parseInt(event.target.value, 10)
+    chrome.send('setDefaultSearchEngine', [modelIndex])
```
@imptrx could you create an action for this?

bsclifton commented:
@imptrx this needs a rebase against master - there's a conflict for the SHA update on `brave-ui`. Can you help me with that?

cezaraugusto commented on components/brave_welcome_ui/components/app.tsx:
```diff
@@ -102,7 +98,12 @@ export class WelcomePage extends React.Component<Props, State> {
+                index={3}
+                currentScreen={this.currentScreen}
+                onClick={this.onClickNext}
+                isDefaultSearchGoogle={true}
```
@bsclifton @imptrx should this be hard-coded `true`? are there an API exposed to check whether or not G is default?

bsclifton replied:
I believe this is for me to populate - will take a look. This would be a matter of detecting whether you are in a region where Google is default (versus Qwant/DDG/etc)

cezaraugusto commented on components/brave_welcome_ui/components/screens/searchBox.tsx:
```diff
@@ -18,11 +19,62 @@ interface Props {
+interface State {
+  searchEngineSelected: boolean,
+  searchEngineIndex: number,
+  searchProviders: Array<any>
```
@imptrx do you think we can improve this type? seems an array of objects with known properties

bsclifton commented:
@cezaraugusto the option for `Choose a search engine` was there before, but I removed. Instead, the engine you are currently using is selected

@rebron @rossmoody do we want to put a blank `Choose a search engine` option back in there? It adds complexity because the search engine list wouldn't be a direct map. Not a big deal, but would just need to add some extra checks

imptrx commented:
@bsclifton there was a more detailed spec that got posted in the sister PR: https://github.com/brave/brave-ui/pull/472#issuecomment-494447686. I can address those details as well as the changes requested

bsclifton commented:
@imptrx awesome, sounds great! 😄 Thanks

imptrx commented:
@bsclifton @cezaraugusto I think this PR is good for another review 😄I've moved the majority of the chrome API logic into redux and added some tests around those. Once everything checks out I'll clean up/squash the commits into chunks

imptrx commented on components/brave_welcome_ui/actions/welcome_actions.ts:
```diff
@@ -15,3 +16,26 @@ export const goToTabRequested = (url: string, target: string) => action(types.GO
+  return (dispatch: Dispatch) => {
+    dispatch(getSearchEngineProvidersStarted())
+
+    // @ts-ignore
```
I couldn't figure out how to type the `sendWithPromise` method 😅- if anyone knows a workaround please reach out

cezaraugusto replied:
@imptrx it should be defined in [brave://resources/js/cr.js](brave://resources/js/cr.js). from there:

```js
  /**
   * A variation of chrome.send, suitable for messages that expect a single
   * response from C++.
   * @param {string} methodName The name of the WebUI handler API.
   * @param {...*} var_args Variable number of arguments to be forwarded to the
   *     C++ call.
   * @return {!Promise}
   */
  function sendWithPromise(methodName, var_args) {
    const args = Array.prototype.slice.call(arguments, 1);
    const promiseResolver = new PromiseResolver();
    const id = methodName + '_' + createUid();
    chromeSendResolverMap[id] = promiseResolver;
    chrome.send(methodName, [id].concat(args));
    return promiseResolver.promise;
  }
```

also what you think of adding an abstraction for `sendWithPromise` instead of having it in here? maybe a file named `welcomeUtils` or similar. not sure if reducer itself have access to the `window` object.

imptrx commented on components/test/brave_welcome_ui/actions/welcome_actions_test.ts:
```diff
@@ -29,4 +29,34 @@ describe('welcome_actions', () => {
+  })
+
+  // TODO: getSearchEngineProviders can be tested via mocking the entire
+  // store. Would require redux-mock-store as dependency
```
@cezaraugusto ☝️thoughts on using https://github.com/dmitry-zaets/redux-mock-store to test the async stuff?

cezaraugusto replied:
👍I've never used it but please add if doable and you think it would help us with testing

bsclifton commented on components/definitions/global.d.ts:
```diff
@@ -13,6 +13,7 @@ declare global {
     loadTimeData: loadTimeData
     cr: {
       define: (name: string, init: () => void) => void
+      sendWithPromise: (method: string, ...args: any[]) => any
```
very nice! I missed this

bsclifton commented:
Quick note- I did see this in the console - not sure if it's a show-stopper (I compiled release mode)
```
[52084:775:0611/131046.411818:ERROR:CONSOLE(496)] "Uncaught TypeError: Cannot read property 'name' of undefined", source: chrome://welcome/brave_welcome.bundle.js (496)
[52084:775:0611/131046.430382:ERROR:CONSOLE(38339)] "The above error occurred in the <SearchEngineBox> component:
    in SearchEngineBox (created by WelcomePage)
    in div (created by styled.div)
    in styled.div (created by WelcomePage)
    in div (created by styled.div)
    in styled.div (created by WelcomePage)
    in div (created by styled.div)
    in styled.div (created by WelcomePage)
    in WelcomePage (created by Connect(WelcomePage))
    in Connect(WelcomePage)
    in ThemeProvider
    in Provider

Consider adding an error boundary to your tree to customize error handling behavior.
Visit https://fb.me/react-error-boundaries to learn more about error boundaries.", source: chrome://welcome/brave_welcome.bundle.js (38339)
```

imptrx commented:
@bsclifton @cezaraugusto @rebron  It looks like when we initially load the welcome(with a blank profile) here is the current behavior:
1. Brave opens the welcome `web-ui` 
2. The page tries to fetch a list of search providers
3. Page crashes because the `window.cr.sendWithPromise('getSearchEnginesList')` doesn't return anything 
4. Doing some other action in Brave will somehow trigger a `searchEngineList` to be set
5. If we load the welcome page again, things will work as expected
So I believe the issue as mentioned earlier is that Brave needs to set a search engine list (along with the current partner default provider for the region) before we try to load any pages. Would that be something we can do on the C++ side?

bsclifton commented:
@imptrx checking it out now...

bsclifton commented:
@imptrx I pushed a fix and verified it works with fresh profiles. I believe a race condition was happening
- request is made
- control is rendered
- error occurred
- (response would have came back here)
- updating props would have triggered a re-render

With this updated logic, it'll render twice. Once before the request finishes and once after 👍

cezaraugusto commented:
waiting on Jenkins before merge
